File: framework/web/src/main/java/top/bettercode/simpleframework/web/SettingsEndpoint.java
package top.bettercode.simpleframework.web;

import java.util.Collections;
import java.util.Map;
import org.springframework.boot.actuate.endpoint.annotation.DeleteOperation;
import org.springframework.boot.actuate.endpoint.annotation.Endpoint;
import org.springframework.boot.actuate.endpoint.annotation.ReadOperation;
import org.springframework.boot.actuate.endpoint.annotation.Selector;
import org.springframework.boot.actuate.endpoint.annotation.WriteOperation;
import org.springframework.lang.Nullable;
import org.springframework.util.StringUtils;
import top.bettercode.lang.property.PropertiesSource;
import top.bettercode.lang.property.Settings;

/**
 * @author Peter Wu
 */
@Endpoint(id = "settings")
public class SettingsEndpoint {

  @WriteOperation
  public Object write(@Selector String baseName, String key, String value) {
    PropertiesSource propertiesSource = Settings.INSTANCE.get(baseName);
    if (propertiesSource == null) {
      return Collections.emptyMap();
    } else {
      propertiesSource.put(key, value);
      return Collections.singletonMap(key, value);
    }
  }

  @DeleteOperation
  public Object delete(@Selector String baseName, String key) {
    PropertiesSource propertiesSource = Settings.INSTANCE.get(baseName);
    if (propertiesSource == null) {
      return Collections.emptyMap();
    } else {
      String remove = propertiesSource.remove(key);
      if (remove == null) {
        Map<String, String> map = propertiesSource.mapOf(key);
        for (String k : map.keySet()) {
          propertiesSource.remove(k);
        }
        return map;
      } else {
        return Collections.singletonMap(key, remove);
      }
    }
  }

  @ReadOperation
  public Object read(@Selector String baseName, @Nullable String key) {
    PropertiesSource propertiesSource = Settings.INSTANCE.get(baseName);
    if (propertiesSource == null) {
      return Collections.emptyMap();
    } else {
      if (StringUtils.hasText(key)) {
        String value = propertiesSource.get(key);
        if (value == null) {
          return propertiesSource.mapOf(key);
        } else {
          return Collections.singletonMap(key, value);
        }
      } else {
        return propertiesSource.all();
      }
    }
  }

}

File: web/src/main/kotlin/top/bettercode/summer/config/SettingsEndpoint.kt
package top.bettercode.summer.config

import org.springframework.boot.actuate.endpoint.annotation.*
import org.springframework.lang.Nullable
import top.bettercode.summer.tools.lang.property.PropertiesSource
import top.bettercode.summer.tools.lang.property.Settings
import java.util.*

/**
 * @author Peter Wu
 */
@Endpoint(id = "settings")
class SettingsEndpoint {

    @WriteOperation
    fun write(@Selector baseName: String?, @Nullable key: String?, @Nullable value: String?): Any {
        val propertiesSource: PropertiesSource? = Settings[baseName]
        return if (propertiesSource == null) {
            emptyMap<Any, Any>()
        } else {
            if (key.isNullOrBlank()) {
                if (value.isNullOrBlank())
                    propertiesSource.load()
                else
                    propertiesSource.load(value)

                propertiesSource.all()
            } else {
                propertiesSource.put(key, value)
                Collections.singletonMap(key, value)
            }
        }
    }

    @DeleteOperation
    fun delete(@Selector baseName: String, key: String?): Any {
        val propertiesSource: PropertiesSource? = Settings[baseName]
        return if (propertiesSource == null) {
            emptyMap<Any, Any>()
        } else {
            val map: MutableMap<String, String?>
            if (!key.isNullOrBlank()) {
                val remove: String? = propertiesSource.remove(key)
                if (Settings.isDicCode(baseName)) {
                    map = propertiesSource.mapOf(key).toMutableMap()
                    val prefix = "$key."
                    for (k in map.keys) {
                        propertiesSource.remove(prefix + k)
                    }
                    val typeKey = "$key|TYPE"
                    val removeType: String? = propertiesSource.remove(typeKey)
                    if (removeType != null) {
                        map["TYPE"] = removeType
                    }
                    if (remove != null) {
                        map[key] = remove
                    }
                } else {
                    map = Collections.singletonMap(key, remove)
                }
            } else {
                map = mutableMapOf()
                map.putAll(propertiesSource.all())
                propertiesSource.clear()
            }
            map
        }
    }

    @ReadOperation
    fun read(@Selector baseName: String, key: String?): Any {
        val propertiesSource: PropertiesSource? = Settings[baseName]
        return if (propertiesSource == null) {
            emptyMap<Any, Any>()
        } else {
            if (!key.isNullOrBlank()) {
                val value: String? = propertiesSource[key]
                val map: MutableMap<String, String?>
                if (Settings.isDicCode(baseName)) {
                    map = propertiesSource.mapOf(key).toMutableMap()
                    val typeKey = "$key|TYPE"
                    val typeValue: String? = propertiesSource[typeKey]
                    if (typeValue != null) {
                        map["TYPE"] = typeValue
                    }
                    if (value != null) {
                        map[key] = value
                    }
                } else {
                    map = Collections.singletonMap(key, value)
                }
                map
            } else {
                propertiesSource.all()
            }
        }
    }
}